### PR TITLE
Ajax: Mitigate possible XSS vulnerability

### PR DESCRIPTION
**Pull request by markelog**
Fixes gh-2432

### CODE REVIEW

**Comment by markelog** [comment]
@timmywil remind me please, did we decided to do it in this or next version? In other words, should we close or land it?

**Comment by mzgol** [comment]
I think we [decided to land it](http://irc.jquery.org/%23jquery-meeting/default_%23jquery-meeting_20150921.log.html#t09:14:49). :)

**Comment by jaubourg** [comment]
Still not a fan of this hack. It looks like old ajax code with tests and special cases hidden deep into the code.

**Comment by mzgol** [comment]
@jaubourg You're thinking from the code perspective, I (others mostly agreed IIRC) just think it's really bad that `$.get(SOME_URL)` can execute code, I think most people doesn't expect it. Methods should be as secure by default as they can be.

**Comment by jaubourg** [comment]
I'm thinking about code perspective because there are better, more consistent ways to handle this than hack into the conversion logic like that. Like providing the options object to converters so that they can handle special cases properly and report sensible errors.

Now, from a behavior point of view, this will effectively prevent valid cross-domain script execution because the xhr transport will provide a text response and will need to use the converter whether the dataType was provided explicitely or not.

**Comment by mzgol** [comment]
OK, I had an impression that you don't want to prevent `$.get(CROSS_DOMAIN_URL)` from executing script and not that you're just against this specific way of achieving that.

**Comment by jaubourg** [comment]
OK, so I've been thinking about this and the best way to handle this is to inhibit auto detection of the script dataType in the context of a cross-domain request which can be achieved by adding the following prefilter before the existing one in src/ajax/script:

```js
// Prevent auto-execution of scripts when no explicit dataType was provided
jQuery.ajaxPrefilter( function( s ) {
	if ( s.crossDomain ) {
		s.contents.script = false;
	}
} );
```

I still think this should be done userland but, anyway, that's the way to achieve this properly.

**Comment by timmywil** [comment]
Thanks @jaubourg. I think that's a good suggestion.

**Comment by reedloden** [comment]
Please see https://github.com/jquery/jquery/issues/2432#issuecomment-199098662

File: test/unit/ajax.js [hunk 1]
```diff
@@ -71,6 +71,54 @@ QUnit.module( "ajax", {
 		};
 	} );
 
+	ajaxTest( "jQuery.ajax() - do not execute js (crossOrigin)", 2, function( assert ) {
```
Comment (FagnerMartinsBrack) [inline test/unit/ajax.js]: Ain't you using `assert.expect` anymore? xD

**Pull request closed without merging**